ALSA push/pop hack breaks Clang 10

error: cannot jump from this goto statement to its label
    goto end;

jump bypasses initialization of variable with __attribute__((cleanup))
    pthread_cleanup_push( &OnExit, stream );

/usr/include/pthread.h:599:36: note: expanded from macro 'pthread_cleanup_push'
    struct __pthread_cleanup_frame __clframe

```
libtool: compile:  clang -c -g -O3 -feliminate-unused-debug-types -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=32 -Wformat -Wformat-security -m64 -fasynchronous-unwind-tables -Wp,-D_REENTRANT -ftree-loop-distribute-patterns -Wl,-z -Wl,now -Wl,-z -Wl,relro -fno-semantic-interposition -ffat-lto-objects -fno-trapping-math -Wl,-sort-common -Wl,--enable-new-dtags -march=westmere -mtune=westmere -mbranches-within-32B-boundaries -w -DPA_LITTLE_ENDIAN -I./include -I./src/common -I./src/os/unix -pthread -DPACKAGE_NAME=\"\" -DPACKAGE_TARNAME=\"\" -DPACKAGE_VERSION=\"\" -DPACKAGE_STRING=\"\" -DPACKAGE_BUGREPORT=\"\" -DPACKAGE_URL=\"\" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_SYS_SOUNDCARD_H=1 -DHAVE_LINUX_SOUNDCARD_H=1 -DSIZEOF_SHORT=2 -DSIZEOF_INT=4 -DSIZEOF_LONG=8 -DHAVE_CLOCK_GETTIME=1 -DHAVE_NANOSLEEP=1 -DPA_USE_ALSA=1 -DPA_USE_OSS=1 src/hostapi/alsa/pa_linux_alsa.c  -fPIC -DPIC -o src/hostapi/alsa/.libs/pa_linux_alsa.o
src/hostapi/alsa/pa_linux_alsa.c:4392:5: error: cannot jump from this goto statement to its label
    goto end;
    ^
src/hostapi/alsa/pa_linux_alsa.c:4217:5: note: jump bypasses initialization of variable with __attribute__((cleanup))
    pthread_cleanup_push( &OnExit, stream );
    ^
/usr/include/pthread.h:599:36: note: expanded from macro 'pthread_cleanup_push'
    struct __pthread_cleanup_frame __clframe                                  \
                                   ^
1 error generated.
make: *** [Makefile:228: src/hostapi/alsa/pa_linux_alsa.lo] Error 1
```

I was able to reproduce this on my Raspberry Pi.

```
sudo apt install clang
clang -c -g -O3 -feliminate-unused-debug-types -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=32 -Wformat -Wformat-security -m64 -fasynchronous-unwind-tables -Wp,-D_REENTRANT -Wl,-z -Wl,now -Wl,-z -Wl,relro -fno-semantic-interposition -ffat-lto-objects -fno-trapping-math -Wl,-sort-common -Wl,--enable-new-dtags -mbranches-within-32B-boundaries -w -DPA_LITTLE_ENDIAN -I./include -I./src/common -I./src/os/unix -pthread -DPACKAGE_NAME=\"\" -DPACKAGE_TARNAME=\"\" -DPACKAGE_VERSION=\"\" -DPACKAGE_STRING=\"\" -DPACKAGE_BUGREPORT=\"\" -DPACKAGE_URL=\"\" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_SYS_SOUNDCARD_H=1 -DHAVE_LINUX_SOUNDCARD_H=1 -DSIZEOF_SHORT=2 -DSIZEOF_INT=4 -DSIZEOF_LONG=8 -DHAVE_CLOCK_GETTIME=1 -DHAVE_NANOSLEEP=1 -DPA_USE_ALSA=1 -DPA_USE_OSS=1 src/hostapi/alsa/pa_linux_alsa.c  -fPIC -DPIC -o src/hostapi/alsa/.libs/pa_linux_alsa.o
```

That led to the error originally reported:
```
src/hostapi/alsa/pa_linux_alsa.c:4471:5: error: cannot jump from this goto statement to its label
    goto end;
```

--- a/src/hostapi/alsa/pa_linux_alsa.c
+++ b/src/hostapi/alsa/pa_linux_alsa.c
@@ -4458,17 +4458,17 @@ static void *CallbackThreadFunc( void *userData )
         }
     }
 
-end:
-    ; /* Hack to fix "label at end of compound statement" error caused by pthread_cleanup_pop(1) macro. */
-    /* Match pthread_cleanup_push */
-    pthread_cleanup_pop( 1 );
-
-    PA_DEBUG(( "%s: Thread %d exiting\n ", __FUNCTION__, pthread_self() ));
-    PaUnixThreading_EXIT( result );
-
+    /* Note that we will not fall into this error label from above.
+     * It is a while(1) loop that only exits using a goto.
+     */
 error:
     PA_DEBUG(( "%s: Thread %d is canceled due to error %d\n ", __FUNCTION__, pthread_self(), result ));
-    goto end;
+
+end:
+    PA_DEBUG(( "%s: Thread %d exiting\n ", __FUNCTION__, pthread_self() ));
+    /* Match pthread_cleanup_push */
+    pthread_cleanup_pop( 1 );
+    PaUnixThreading_EXIT( result );
 }
 
 /* Blocking interface */

